aws_lambda_builders/workflows/rust_cargo/actions.py

```diff
@@ -0,0 +1,216 @@
+
+    def popen(self, command, stdout=None, stderr=None, env=None, cwd=None):
+        return subprocess.Popen(command, stdout=stdout, stderr=stderr, env=env, cwd=cwd)
+
+    def copyfile(self, source, destination):
+        shutil.copyfile(source, destination)
+
+    def makedirs(self, path):
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+
+def parse_handler(handler):
+    """
+    Parse package and binary from handler name
+
+    If the handler contains a period, assume `package.bin_name`
+    otherwise assume common case where bin_name is the same as the package
+    """
+    spec = handler.split(".", 1)
+    if len(spec) == 1:
+        return (spec[0], spec[0])
+    else:
+        return (spec[0], spec[1])
+
+
+class BuildAction(BaseAction):
+    NAME = "CargoBuild"
+    DESCRIPTION = "Building the project using Cargo"
+    PURPOSE = Purpose.COMPILE_SOURCE
+
+    def __init__(self, source_dir, handler, binaries, platform, mode, osutils=OSUtils()):
+        """
+        Build the a rust executable
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type handler: str
+        :param handler:
+            Handler name in `package.bin_name` or `bin_name` format
+
+        :type binaries: dict
+        :param binaries:
+            Resolved path dependencies
+
+        :type platform: string
+        :param platform:
+            Platform builder is being run on
+
+        :type mode: str
+        :param mode:
+            Mode the build should produce
+
+        :type osutils: object
+        :param osutils:
+            Optional, External IO utils
+        """
+        self.source_dir = source_dir
+        self.handler = handler
+        self.mode = mode
+        self.binaries = binaries
+        self.platform = platform
+        self.osutils = osutils
+
+    def cargo_metadata(self):
+        p = self.osutils.popen(
+            ["cargo", "metadata", "--no-deps", "--format-version=1"],
+            stderr=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            env=os.environ.copy(),
+            cwd=self.source_dir,
+        )
+        out, err = p.communicate()
+        if p.returncode != 0:
+            raise BuilderError(message=err.decode("utf8").strip())
+        return json.loads(out)
+
+    def build_command(self, package):
+        cmd = [self.binaries["cargo"].binary_path, "build", "-p", package]
+        if self.mode != BuildMode.DEBUG:
+            cmd.append("--release")
+        if self.platform.lower() != "linux":
+            cmd.extend(["--target", "x86_64-unknown-linux-musl"])
+        return cmd
+
+    def resolve_binary(self, cargo_meta):
+        """
+        Interrogate cargo metadata to resolve a handler function
+
+        :type cargo_meta: dict
+        :param cargo_meta:
+            Build metadata emitted by cargo
+        """
+        (package, binary) = parse_handler(self.handler)
+        exists = any(
```

Comment:
I realize this expression is kind of dense to parse but I didn't know if there was a more pythonic way of expressing this beyond a series of `maps` and `anys`

Reply:
hmm.. Let me think through this and parse it out. This is hard to understand in the current state.

Reply:
You can use the tests for this to guide your understanding.